### PR TITLE
Respect slot size given by KDS

### PR DESCRIPTION
**Pull request by chienwei-lan**

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c [hunk 1]
```diff
@@ -15,7 +15,7 @@
  */
 #include "xclbin.h"
 #include "ps_kernel.h"
-
+
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c:18]: 'Tab' in an empty line.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c [hunk 2]
```diff
@@ -1079,21 +1073,35 @@ static int ert_cfg_cmd(struct xocl_ert_user *ert_user, struct ert_user_command *
 		ERTUSER_ERR(ert_user, "should not have zeroed value of cq_size=%lld, slot_size=%d",
 		    ert_user->cq_range, cfg->slot_size);
 		return -EINVAL;
+	} else if (cfg->slot_size & 0x1F) {
+		ERTUSER_ERR(ert_user, "slot_size should be 32-bit aligned, slot_size=%d",
+		   cfg->slot_size);
+		return -EINVAL;
 	}
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c:1080]: The slot_size is in bytes. "cfg->slot_size & 0x1F" here means it is 32 bytes aligned. Right?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c [hunk 3]
```diff
@@ -19,6 +19,7 @@
 #include "kds_client.h"
 
 #define	ERT_MAX_SLOTS		128
+#define	ERT_MIN_SLOT_SIZE	512
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c:22]: I would suggest do not define this macro.
The minimum slot size is defined by "ert_user->cq_range / ERT_MAX_SLOTS". If there is a shell has different cq_range, use this formula would still work.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c [hunk 4]
```diff
@@ -1079,21 +1073,35 @@ static int ert_cfg_cmd(struct xocl_ert_user *ert_user, struct ert_user_command *
...
+	 *  2. the slot num max: 128
+	 *  3. slot size at least 512 bytes 
+	 */
 
-	if (ert_poll)
-		// Adjust slot size for ert poll mode
-		ert_num_slots = MAX_CUS;
+	if (ert_num_slots > MAX_CUS)
+		// Adjust slot number to 128 maximum 
+		slot_size = ert_user->cq_range / MAX_CUS;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c:1096]: Please use ERT_MAX_SLOTS.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c [hunk 5]
```diff
@@ -1079,21 +1073,35 @@ static int ert_cfg_cmd(struct xocl_ert_user *ert_user, struct ert_user_command *
...
+		ERTUSER_ERR(ert_user, "slot_size should be 32-bit aligned, slot_size=%d",
+		   cfg->slot_size);
+		return -EINVAL;
 	}
 
-	ert_num_slots = ert_user->cq_range / cfg->slot_size;
+	slot_size = cfg->slot_size;
+
+	if (slot_size < ERT_MIN_SLOT_SIZE)
+		slot_size = ERT_MIN_SLOT_SIZE;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c:1085]: See my comment about this macro.

**Review by mamin506 — CHANGES_REQUESTED** [review]
See my comments.

**Review by mamin506 — CHANGES_REQUESTED** [review]
Minor change.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c [hunk 6]
```diff
@@ -1079,21 +1072,35 @@ static int ert_cfg_cmd(struct xocl_ert_user *ert_user, struct ert_user_command *
 		ERTUSER_ERR(ert_user, "should not have zeroed value of cq_size=%lld, slot_size=%d",
 		    ert_user->cq_range, cfg->slot_size);
 		return -EINVAL;
+	} else if (!IS_ALIGNED(cfg->slot_size, 4)) {
+		ERTUSER_ERR(ert_user, "slot_size should be 32-bit aligned, slot_size=%d",
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c:1076]: Please update the error message as well, "slot_size should be 4 bytes aligned".

**Comment by dayeh-xilinx** [comment]
retest this please. Pipeline problem with Docker image downloading should be fixed now.

**Review by mamin506 — APPROVED** [review]

**Pull request merged**